src/main/java/frc/trigon/robot/poseestimation/robotposesources/RobotPoseSource.java

```diff
@@ -76,11 +66,64 @@ public double getLastResultTimestamp() {
         return inputs.lastResultTimestamp;
     }
 
-    private Pose2d getUnCachedRobotPose() {
-        final Pose3d cameraPose = doubleArrayToPose3d(inputs.cameraPose);
+
+    /**
+     * Calculates the standard deviations of the estimated robot pose.
+     *
+     * @return the standard deviations for the pose estimation strategy used
+     */
+    public Matrix<N3, N1> calculateStdDevs() {
+        if (isWithinTagRange())
+            return calculateSolvePNPStdDevs();
+        return calculateAssumedHeadingStdDevs();
+    }
+
+    /**
+     * If the robot is close enough to the tag, it gets the estimated solve PNP pose.
+     * If it's too far, it assumes the robot's heading and calculates its position from there.
+     * Assuming the robot's heading is more robust, but using solve PNP resets the robot's heading.
```

Comment:
Assuming the robot's heading is more robust, but won't fix current wrong heading. To fix this, we're using solve PNP to reset the robot's heading, when we're close enough for a very accurate result.